cpp/src/arrow/compute/exec/options.h

```diff
@@ -397,23 +397,33 @@ class ARROW_EXPORT HashJoinNodeOptions : public ExecNodeOptions {
 /// This node will output one row for each row in the left table.
 class ARROW_EXPORT AsofJoinNodeOptions : public ExecNodeOptions {
  public:
-  AsofJoinNodeOptions(FieldRef on_key, FieldRef by_key, int64_t tolerance)
-      : on_key(std::move(on_key)), by_key(std::move(by_key)), tolerance(tolerance) {}
+  AsofJoinNodeOptions(FieldRef on_key, const FieldRef& by_key, int64_t tolerance)
+      : on_key(std::move(on_key)), by_key(), tolerance(tolerance) {
+    this->by_key.push_back(std::move(by_key));
+  }
+
+  AsofJoinNodeOptions(FieldRef on_key, std::vector<FieldRef> by_key, int64_t tolerance)
+      : on_key(std::move(on_key)), by_key(by_key), tolerance(tolerance) {}
+
+  // resolves ambiguity between previous constructors when initializer list is given
+  AsofJoinNodeOptions(FieldRef on_key, std::initializer_list<FieldRef> by_key,
```

Comment:
What is the constructor used for?

Reply:
The other two constructors, one accepting a `FieldRef` parameter and the other a `std::vector<FieldRef>` parameter, cannot be disambiguated when one writes something like `AsOfJoinOptions("time", {"key1", "key2"}, tolerance)` because `` fits both parameter types `{"key1", "key2"}` with a single implicit conversion. However, once a constructor with `std::initializer_list<FieldRef>` is added, this is the constructor the compiler will pick, since it involves no implicit conversion.

Reply:
Can we only allow passing as `std::vector<FieldRef>`?

Reply:
(I think this is less code this way and not a big issue to always use {"key1"} for a single by key IMO

Reply:
> Can we only allow passing as `std::vector<FieldRef>`?

I thought about it but this would break backwards-compatibility. Unless you can confirm `AsofJoin` is experimental and allowed to break?

Reply:
> Thanks @rtpsw for the benchmark numbers. Do you think the regression are caused by key hashing?

No, because the benchmark was written prior to this feature and must invoke the fast path. I suspect it is overhead in hot-spots such as `GetLatestKey`. I'm looking into it.

Reply:
[Fixed](https://github.com/apache/arrow/pull/13880#issuecomment-1235468859).